api/src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java

```diff
@@ -92,11 +92,33 @@ public Component associateCollectorToComponent(ObjectId componentId, List<Object
             return null;
         }
 
-        Component component = componentRepository.findOne(componentId);
+        com.capitalone.dashboard.model.Component component = componentRepository.findOne(componentId);
+        //we can not assume what collector item is added, what is removed etc so, we will
+        //refresh the association. First disable all collector items, then remove all and re-add
+
+        //First: disable all collectorItems of the Collector TYPEs that came in with the request.
+        //Second: remove all the collectorItem association of the Collector Type  that came in
+        List<CollectorType> incomingTypes = new ArrayList<>();
         for (ObjectId collectorItemId : collectorItemIds) {
             CollectorItem collectorItem = collectorItemRepository.findOne(collectorItemId);
             Collector collector = collectorRepository.findOne(collectorItem.getCollectorId());
+            if (!incomingTypes.contains(collector.getCollectorType())) {
+                incomingTypes.add(collector.getCollectorType());
+                List<CollectorItem> cItems = component.getCollectorItems(collector.getCollectorType());
+                if (!CollectionUtils.isEmpty(cItems)) {
+                    for (CollectorItem ci : cItems) {
+                        ci.setEnabled(false);
+                        collectorItemRepository.save(ci);
+                    }
+                }
+                component.getCollectorItems().remove(collector.getCollectorType());
+            }
+        }
 
+        //Last step: add collector items that came in
+        for (ObjectId collectorItemId : collectorItemIds) {
```

Comment:
They can possibly be combined - but I would rather have two distinct clean remove/add steps and not combine them - hard to debug and understand. Most cases will have only one to loop.